Speaker/Speaker.swift

```diff
@@ -0,0 +1,161 @@
+
+        guard createError == noErr else {
+            assertionFailure("Error creating audio queue")
+            return nil
+        }
+
+        let isRunningListenerError = AudioQueueAddPropertyListener(audioQueue, kAudioQueueProperty_IsRunning, Self.startedListener, bridge(obj: self))
+
+        guard isRunningListenerError == noErr else {
+            assertionFailure("Error adding listener to audio queue")
+            return nil
+        }
+
+        for _ in 0..<Self.numberOfBuffers {
+            var audioQueueBuffer: AudioQueueBufferRef?
+            let error = AudioQueueAllocateBuffer(audioQueue, bufferSize, &audioQueueBuffer)
+
+            guard error == noErr else {
+                assertionFailure("Failed to create buffer")
+                return nil
+            }
+
+            guard let buffer = audioQueueBuffer else {
+                assertionFailure("Buffer is nil")
+                return nil
+            }
+
+            buffers.append(buffer)
+        }
+
+        let startError = AudioQueueStart(audioQueue, nil)
+
+        guard startError == noErr else {
+            assertionFailure("Failed to start")
+            return nil
+        }
+    }
+
+    deinit {
+        let error = AudioQueueDispose(audioQueue, true)
+        assert(error == noErr, "Error disposing audio queue")
+    }
+
+    private var audioQueue: AudioQueueRef!
+    private var buffers: [AudioQueueBufferRef] = []
+
+    private static let numberOfBuffers = 300
+    private let bufferSize: UInt32
+
+    private static let startedListener: AudioQueuePropertyListenerProc = { userData, audioQueue, propertyID in
+
+        guard propertyID == kAudioQueueProperty_IsRunning else {
+            assertionFailure("Wrong property")
+            return
+        }
+
+        var isRunning: UInt32 = 0
+        var size = UInt32(MemoryLayout.size(ofValue: UInt32.self))
+
+        let error = AudioQueueGetProperty(audioQueue, kAudioQueueProperty_IsRunning, &isRunning, &size)
+
+        guard error == noErr else {
+            assertionFailure("Failed to get prop isRunning")
+            return
+        }
+
+        guard let userData = userData else {
+            assertionFailure("UserData is nil")
+            return
+        }
+
+        let speaker: Speaker = bridge(ptr: userData)
+
+        if isRunning > 0 {
+            speaker.buffers.forEach { buffer in
+                Speaker.audioQueueOutputCallback(userData, audioQueue, buffer)
+            }
+        }
+    }
+
+    private static let audioQueueOutputCallback: AudioQueueOutputCallback = { userData, audioQueue, buffer in
+
+        guard let userData = userData else {
+            assertionFailure("UserData is nil")
+            return
+        }
+
+        let speaker: Speaker = bridge(ptr: userData)
+
+        // Fill buffer with silence
+        memset(buffer.pointee.mAudioData, 0, Int(speaker.bufferSize))
```

Comment:
I don't think this is necessary as the buffer should be filled completely by the rustboy method. Also it's being copied across in the next block anyway.

Reply:
This is for if you start the speaker before the delegate is set. As soon as the delegate is set it will start filling it with data from it.

Reply:
I moved it behind an else to make sure we aren't doing it when we don't need to